src/rpc/rpc.cc

```diff
@@ -76,14 +76,57 @@ temporary_buffer<char>& snd_buf::front() {
 
 namespace internal {
 
+void snd_buf_batched_queue::enqueue(snd_buf* buf) {
+    _queue.push_back(buf);
+    if (_process_fut.available()) {
+        _process_fut = process_loop();
+    }
+}
+
+future<> snd_buf_batched_queue::process_loop() {
+    return seastar::do_until([this] { return _queue.empty(); }, [this] {
+        _cur_batch.clear();
+        std::swap(_cur_batch, _queue);
+        _cur_batch_pos = _cur_batch.begin();
+        return smp::submit_to(_processing_shard, [this] {
+            return seastar::do_until([this] { return _cur_batch_pos == _cur_batch.end(); }, [this] {
+                auto* buf = *_cur_batch_pos;
+                ++_cur_batch_pos;
+                return _process_func(buf);
```

Comment:
Why not co-routine and then you can just loop with `for (auto i : _cur_batch)`?

Reply:
do_until allocates a task only if its subordinate func does not return a ready future.